pkg/twcc/twcc.go

```diff
@@ -33,68 +37,118 @@ type Recorder struct {
 // feedback packets.
 func NewRecorder(senderSSRC uint32) *Recorder {
 	return &Recorder{
-		receivedPackets: []pktInfo{},
-		senderSSRC:      senderSSRC,
+		senderSSRC: senderSSRC,
 	}
 }
 
 // Record marks a packet with mediaSSRC and a transport wide sequence number sequenceNumber as received at arrivalTime.
 func (r *Recorder) Record(mediaSSRC uint32, sequenceNumber uint16, arrivalTime int64) {
 	r.mediaSSRC = mediaSSRC
-	if sequenceNumber < 0x0fff && (r.lastSequenceNumber&0xffff) > 0xf000 {
-		r.cycles += 1 << 16
+
+	// "Unwrap" the sequence number to get a monotonically increasing sequence number that
+	// won't wrap around after math.MaxUint16.
+	unwrappedSN := r.sequenceUnwrapper.Unwrap(sequenceNumber)
+	r.maybeCullOldPackets(unwrappedSN, arrivalTime)
+	if r.startSequenceNumber == nil || unwrappedSN < *r.startSequenceNumber {
+		r.startSequenceNumber = &unwrappedSN
 	}
-	r.receivedPackets = insertSorted(r.receivedPackets, pktInfo{
-		sequenceNumber: r.cycles | uint32(sequenceNumber),
-		arrivalTime:    arrivalTime,
-	})
-	r.lastSequenceNumber = sequenceNumber
-}
 
-func insertSorted(list []pktInfo, element pktInfo) []pktInfo {
-	if len(list) == 0 {
-		return append(list, element)
+	// We are only interested in the first time a packet is received.
+	if r.arrivalTimeMap.HasReceived(unwrappedSN) {
+		return
 	}
-	for i := len(list) - 1; i >= 0; i-- {
-		if list[i].sequenceNumber < element.sequenceNumber {
-			list = append(list, pktInfo{})
-			copy(list[i+2:], list[i+1:])
-			list[i+1] = element
-			return list
-		}
-		if list[i].sequenceNumber == element.sequenceNumber {
-			list[i] = element
-			return list
-		}
+
+	r.arrivalTimeMap.AddPacket(unwrappedSN, arrivalTime)
+
+	// Limit the range of sequence numbers to send feedback for.
+	if *r.startSequenceNumber < r.arrivalTimeMap.BeginSequenceNumber() {
+		sn := r.arrivalTimeMap.BeginSequenceNumber()
+		r.startSequenceNumber = &sn
+	}
```

Comment:
Can this check be moved to when building feedback? If an out-of-order comes, it could again get back to some value and again reset here because it is before the beginning sequence number. Wondering if it can be done just when building feedback.

While I do not see any issues in the current implemtation, wondering if the implementation can be done without exposing the `BeginSequenceNumber()`, `EndSequenceNumber()` API from the arrival time map. Those could be a bit risky, i. e. if users do things like calculating size based on those values, etc, it could lead to unintentional errors. Again, nothing of that sort in this PR, but just being cautious. Will leave a few more comments along those lines below.

Thought about this when I was reading `EraseTo` function which resets those numbers (a bunch of other places change them also, but that function caught my eye as begin and end are set to same to indicate empty, but was thinking of chances of users of that arrival map module holding on to stale/outdated begin/end). BTW, did not see use of `EraseTo` API except in tests.

Along those lines would actually be good if this module can eliminate even holding the start sequence. I have not thought about if that can also be rolled into the map (i. e. the map is always source of truth), but maybe worth looking at.

Reply:
`packetArrivalTimeMap` itself is unexported, so I don't believe users will be able to call those methods. I think you raise some good points, but since they are private APIs, and consistent with libwebrtc, I think I'll leave as-is